Automattic/jetpack
# Widgets: Fix the Simple Payments widget in the widgets.php page

Fixes https://github.com/Automattic/wp-calypso/issues/25576

The newly introduced Simple Payments widget heavily relies on the Customizer JS API and therefore it doesn't work in the Widgets (`widgets.php`) page.

This PR is an attempt to make it work in both pages flawlessly.

My initial attempt consisted in revers engineer the Customizer itself, to figure out when and why all widget events are triggered, in order to emulate them in the Widgets page.
This was sort of possible, but required an amount of hacks and workarounds several orders of magnitude larger than what I'm comfortable with.

This is the second attempt, and it's completely independent from the Customizer JS API.
It relies entirely on the event delegation, which is easy and performant enough to pull off with jQuery.
I haven't modified the existing `form.php` and `customizer.js` files, but instead opted to write two new files from scratch in order to keep both version always open side by side, trying to stay as true to the original code as possible, while also rewriting it and learning from it.

The idea behind this attempt is that we observe the widgets container, and only react when one of the buttons in a Simple Payments widget is clicked. From there, that widget becomes the context where everything happens.

Clicking on "Add New" will simply open an empty form.
Clicking on "Edit Selected" will fetch via AJAX the details of the selected product to populate the form.
Saving will simply save via AJAX whatever is in the form, and update the products selector accordingly.
Deleting will delete the product and update the products selector.
Closing the form will simply hide and clear it.

Bar corner cases and polish, this works fine in both the Widgets page and the Customizer.

There is one gotcha that I haven't got the time to study because I caught it too late before my AFK time.
When adding a new product or editing an existing one, the Customizer does not refresh accordingly until the product is saved.
This happens because the selective refresh relies on the values returned by the `WP_Widget::update()` method. In this case, the only values actually needed for the widget are the title and a product ID.
All the product details (name, price, image, etc.) are saved via AJAX and stored in their own posts of CPT `jp_pay_product`.
I know for a fact that @rodrigoi solved this issue, but for my poor understanding of the selective refresh magic, I can't seem to figure out exactly how.

Since my AFK is near, I'm leaving this PR here hoping it will be of some inspiration to fix that nasty issue.

#### Changes proposed in this Pull Request:

* WIP

#### Testing instructions:

* On a Pro JP site, open both the Widgets page and the Customizer.
* Add multiple Simple Payments widgets to a sidebar.
* Try to add new products, and edit and deleting existing ones.
* Make sure all the operations work as expected.
* Make sure the Customizer preview updates accordingly to the edit (note: as of now, products are only updated when saving them).
